### PR TITLE
Implement volume resize for cinder

### PR DESCRIPTION
**Pull request by NickrenREN**
**What this PR does / why we need it**:
resize for cinder
xref: [resize proposal](https://github.com/kubernetes/community/pull/657)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: xref https://github.com/kubernetes/community/pull/657
Follow up: #49727

**Special notes for your reviewer**:

**Release note**:
```release-note
Implement volume resize for cinder
```

wip, assign to myself first

/assign @NickrenREN

### CODE REVIEW

**Comment by NickrenREN** [comment]
cc @gnufied , wip, need to add fs resize

**Comment by huzhengchuan** [comment]
cc @huzhengchuan

**Comment by huzhengchuan** [comment]
Good work!
I have test the pr using openstack with cloud privide. the openstack version is mitaka. when modify the storage of pvc, the volume of cinder is resize. 
when fs resize for volume is implements, I will test again.

**Comment by NickrenREN** [comment]
@huzhengchuan  thanks for doing this. When fs resize code gets merged, i will ping you.
cc @gnufied

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 1]
```diff
@@ -193,6 +195,36 @@ func (volumes *VolumesV2) deleteVolume(volumeID string) error {
...
+func (volumes *VolumesV1) expandVolume(volumeID string, newSize int) error {
+	startTime := time.Now()
+	create_opts := volumeexpand.ExtendSizeOpts{
+		NewSize: newSize,
+	}
+	err := volumeexpand.ExtendSize(volumes.blockstorage, volumeID, create_opts).ExtractErr()
+	timeTaken := time.Since(startTime).Seconds()
+	recordOpenstackOperationMetric("expand_volume", timeTaken, err)
+	if err != nil {
+		glog.Errorf("Cannot expand volume %s: %v", volumeID, err)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Don't log-and-return errors, just return them.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 2]
```diff
@@ -193,6 +195,36 @@ func (volumes *VolumesV2) deleteVolume(volumeID string) error {
...
+func (volumes *VolumesV2) expandVolume(volumeID string, newSize int) error {
+	startTime := time.Now()
+	create_opts := volumeexpand.ExtendSizeOpts{
+		NewSize: newSize,
+	}
+	err := volumeexpand.ExtendSize(volumes.blockstorage, volumeID, create_opts).ExtractErr()
+	timeTaken := time.Since(startTime).Seconds()
+	recordOpenstackOperationMetric("expand_volume", timeTaken, err)
+	if err != nil {
+		glog.Errorf("Cannot expand volume %s: %v", volumeID, err)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Another log-and-return

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 3]
```diff
@@ -288,6 +320,32 @@ func (os *OpenStack) DetachDisk(instanceID, volumeID string) error {
...
 
+// ExpandVolume expands the size of specific cinder volume (in GiB)
+func (os *OpenStack) ExpandVolume(volumeID string, newSize int) error {
+	volume, err := os.getVolume(volumeID)
+	if err != nil {
+		return err
+	}
+	if volume.Status != VolumeAvailableStatus {
+		// cinder volume can not be expanded if its status is not available
+		glog.V(2).Infof("the status of volume: %s is not available", volume.ID)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Don't log-and-return.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 4]
```diff
@@ -288,6 +320,32 @@ func (os *OpenStack) DetachDisk(instanceID, volumeID string) error {
...
+		return fmt.Errorf("volume status is not available")
+	}
+
+	volumes, err := os.volumeService("")
+	if err != nil || volumes == nil {
+		glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: You know what I'm going to say at this point ;)  Also, this error doesn't actually include `err`.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 5]
```diff
@@ -288,6 +320,32 @@ func (os *OpenStack) DetachDisk(instanceID, volumeID string) error {
...
+		return err
+	}
+
+	newSizeGb := newSize / 1024 / 1024 / 1024
+	err = volumes.expandVolume(volumeID, newSizeGb)
+	if err != nil {
+		glog.Errorf("Cannot expand volume %s: %v", volumeID, err)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: ditto log-and-return

File: pkg/volume/cinder/cinder.go [hunk 6]
```diff
@@ -227,6 +228,34 @@ func (plugin *cinderPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*
 	return volume.NewSpecFromVolume(cinderVolume), nil
 }
 
+var _ volume.ExpandableVolumePlugin = &cinderPlugin{}
+
+func (plugin *cinderPlugin) ExpandVolumeDevice(spec *volume.Spec, newSize resource.Quantity, oldSize resource.Quantity) (resource.Quantity, error) {
+	cinder, _, err := getVolumeSource(spec)
+	if err != nil {
+		glog.Errorf("error when getting volume source : %v", err)
```
Comment (anguslees) [inline pkg/volume/cinder/cinder.go]: more log-and-returns all the way through here.  I won't bother calling them out individually.

**Review by anguslees — CHANGES_REQUESTED** [review]
Nice, lgtm on all the important bits.

**Comment by NickrenREN** [comment]
Thanks for your review @anguslees  i will delete the log

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 1 above)
Comment (NickrenREN) [reply pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Yeah, i will delete the log

Comment (NickrenREN) [reply pkg/cloudprovider/providers/openstack/openstack_volumes.go]: done

**Comment by NickrenREN** [comment]
@anguslees i have deleted all the err logs and just return them now. PTAL, thanks

**Review by anguslees — APPROVED** [review]

**Comment by NickrenREN** [comment]
@anguslees  I remove the `WIP` and update the bazel file.

File: pkg/volume/cinder/cinder.go [hunk 7]
```diff
@@ -227,6 +228,31 @@ func (plugin *cinderPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*
...
+	cinder, _, err := getVolumeSource(spec)
+	if err != nil {
+		return oldSize, err
+	}
+	cloud, err := plugin.getCloudProvider()
+	if err != nil {
+		return oldSize, err
+	}
+
+	err = cloud.ExpandVolume(cinder.VolumeID, int(newSize.Value()))
```
Comment (gnufied) [inline pkg/volume/cinder/cinder.go]: `ExpandVolume`  ideally should take `resource.Quantity` as input and return a `resource.Quantity` as return value which represents new size.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 8]
```diff
@@ -268,6 +294,27 @@ func (os *OpenStack) DetachDisk(instanceID, volumeID string) error {
...
+		// cinder volume can not be expanded if its status is not available
+		return fmt.Errorf("volume status is not available")
+	}
+
+	volumes, err := os.volumeService("")
+	if err != nil {
+		return err
+	}
+
+	newSizeGb := newSize / 1024 / 1024 / 1024
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: In stead of dividing by 1024, can we use `volume.RoundUpSize(expansionSize, 1024*1024*1024)`  ? Also as I commented below - most volume types after resizing also return new size. We should stick to that convention.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 9]
```diff
@@ -268,6 +294,27 @@ func (os *OpenStack) DetachDisk(instanceID, volumeID string) error {
 	return nil
 }
 
+// ExpandVolume expands the size of specific cinder volume (in GiB)
+func (os *OpenStack) ExpandVolume(volumeID string, newSize int) error {
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Calling this function should be idempotent. I am wondering if there is a way we can describe the disk and check the size before deciding if resizing is required.

**Review by gnufied — COMMENTED** [review]
Mostly looks okay. Left some minor comments.

**Comment by gnufied** [comment]
Also, you will have to update resize admission plugin to allow resizing of Cinder volumes.

File: pkg/volume/cinder/cinder.go (hunk 7 above)
Comment (NickrenREN) [reply pkg/volume/cinder/cinder.go]: I just try to keep this consistent with `CreateVolume` using `int` input. And also `gophercloud` lib will just return err if sth is wrong and nil if everything is ok.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 8 above)
Comment (NickrenREN) [reply pkg/cloudprovider/providers/openstack/openstack_volumes.go]: yeah, will do

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 9 above)
Comment (NickrenREN) [reply pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Done

File: pkg/volume/cinder/cinder.go (hunk 7 above)
Comment (NickrenREN) [reply pkg/volume/cinder/cinder.go]: I changed that in the end

**Comment by NickrenREN** [comment]
@gnufied  Addressed your comments, ptal, thanks

**Comment by jsafrane** [comment]
/assign @thockin 
for approval

**Comment by NickrenREN** [comment]
Thanks @jsafrane  could you help add 1.9 milestone for this one too ? 
https://github.com/kubernetes/kubernetes/pull/52767

**Pull request merged**